onnxruntime/core/providers/mkldnn/nn/batch_norm.cc

```diff
@@ -0,0 +1,302 @@
+    : src_dims(src_dims_mkl),
+    scale_dims(scale_dims_mkl),
+    b_dims(b_dims_mkl),
+    mean_dims(mean_dims_mkl),
+    var_dims(var_dims_mkl),
+    dst_dims(dst_dims_mkl),
+    epsilon(eps),
+    num_dimensions(dimensions) {}
+
+  // Used as the key for BatchNorm Primitive Reuse Pool.
+  std::string ToString() const {
+    std::string key;
+    key.reserve(128);
+    key.append("BatchNorm_");
+    AddDimsToKey(key, src_dims);
+    AddDimsToKey(key, scale_dims);
+    AddDimsToKey(key, b_dims);
+    AddDimsToKey(key, mean_dims);
+    AddDimsToKey(key, var_dims);
+    AddDimsToKey(key, dst_dims);
+    return key;
+  }
+};
+
+template <typename T>
+class BatchNormPrimitive final : public PrimitiveBase {
+ public:
+  explicit BatchNormPrimitive(const BatchNormParams& params)
+      : cpu_engine_(GetEngine()) {
+    context_.stream.reset(new mkldnn::stream(mkldnn::stream::kind::eager));
+    if (context_.batchnorm_fwd == nullptr) {
+      Initialize(params);
+    }
+  }
+
+  ~BatchNormPrimitive() = default;
+
+  void Compute(const T* src_data, const T* scale_data, const T* b_data, 
+    const T* mean_data, const T* var_data, const T* dst_data, 
+    int scale_dims_channels) {
+    context_.src_mem->set_data_handle(
+      static_cast<void*>(const_cast<T*>(src_data)));
+    context_.mean_mem->set_data_handle(
+      static_cast<void*>(const_cast<T*>(mean_data)));
+    context_.var_mem->set_data_handle(
+      static_cast<void*>(const_cast<T*>(var_data)));
+    context_.dst_mem->set_data_handle(
+      static_cast<void*>(const_cast<T*>(dst_data)));
+
+    T* scaleShift_buf = static_cast<T*>(context_.scale_shift_mem->get_data_handle());
+
+    size_t src_bytes = sizeof(T) * scale_dims_channels;
+    size_t dst_bytes = sizeof(T) * scale_dims_channels;
+
+    MEMCPY_S(scaleShift_buf, scale_data, src_bytes, dst_bytes);
+    MEMCPY_S(&scaleShift_buf[scale_dims_channels], b_data, src_bytes, dst_bytes);
+    context_.stream->submit(context_.net);
+    return;
+  }
+
+  std::unique_ptr<mkldnn::convolution_forward::primitive_desc>
+  GetPrimitiveDesc() const {
+    return context_.conv_fwd_pd;
+  }
+
+ private:
+  struct BatchNormContext {
+    std::unique_ptr<mkldnn::memory> src_mem;
+    std::unique_ptr<mkldnn::memory> scale_shift_mem;
+    std::unique_ptr<mkldnn::memory> mean_mem;
+    std::unique_ptr<mkldnn::memory> var_mem;
+    std::unique_ptr<mkldnn::memory> dst_mem;
+
+    std::unique_ptr<mkldnn::memory::desc> src_md;
+    std::unique_ptr<mkldnn::memory::desc> scale_shift_md;
+    std::unique_ptr<mkldnn::memory::desc> mean_md;
+    std::unique_ptr<mkldnn::memory::desc> var_md;
+    std::unique_ptr<mkldnn::memory::desc> dst_md;
+
+    std::unique_ptr<mkldnn::batch_normalization_forward::desc> batchnorm_fwd;
+    std::unique_ptr<mkldnn::batch_normalization_forward::primitive_desc> 
+      batchnorm_fwd_pd;
+
+    std::unique_ptr<mkldnn::stream> stream;
+    std::vector<mkldnn::primitive> net;
+
+    BatchNormContext()
+      : src_mem(nullptr),
```

Comment:
unique_ptrs are nullptrs by default. there is no need to initialize them.

Reply:
removed constructor in Sum and BatchNorm.